<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title>a Real Promise case</title>
    <meta name="viewport" content="width=device-width, initial-scale=1">
</head>
<body>
    <script>
        function getJson(url) {
            return new Promise((resolve, reject) => {
                const request = new XMLHttpRequest();

                request.open("GET", url);

                request.onload = function() {
                    try {
                        console.log(this.status);
                        if(this.status === 200 || this.status ===304) {
                            resolve(JSON.parse(this.response));
                        } else {
                            reject(this.status + " " + this.statusText);
                        }
                    } catch(e) {
                        reject(e.message);
                    }
                };

                request.onerror = function() {
                    reject(this.status + " " + this.statusText);
                };

                request.send();
            });
        }

        getJson("/api/issues")
        .then(issues => console.log(issues))
        .catch(err => console.log(err));

        async(function*() {
            try {
                const issues1 = yield getJson("api/issues");
                const issues2 = yield getJson("api/issues");
                const issues3 = yield getJson("api/issues");
            } catch(error) {
                console.log(error);
            }
        });

        function async(generator) {
            const iterator = generator();

            function handle(iteratorResult) {
                if(iteratorResult.done) { return; }

                const iteratorValue = iteratorResult.value;

                if(iteratorValue instanceof Promise) {
                    iteratorValue
                    .then(issues => {
                        console.log(issues);
                        handle(iterator.next())
                    })
                    .catch(err => iterator.throw(err));
                }
            }

            try {
                handle(iterator.next());
            } catch(err){
                iterator.throw(error);
            }
        }
    </script>
    
</body>
</html>